src/realm_dart_collections.cpp

```diff
@@ -19,84 +19,82 @@
 #include "realm.h"
 #include "dart_api_dl.h"
 #include "realm_dart_collections.h"
+#include <stdio.h>
 
-struct Scope
-{
-    Scope()
-    {
+struct HandleScope {
+    HandleScope() {
         Dart_EnterScope_DL();
     }
-    ~Scope()
-    {
+
+    ~HandleScope() {
         Dart_ExitScope_DL();
     }
 };
 
-class Callback
-{
-    Dart_WeakPersistentHandle handle_;
-    realm_dart_on_collection_change_func_t callback_;
-
-    static void finalize_(void *isolate_callback_data, void *peer)
-    {
-        auto &callback = *reinterpret_cast<Callback *>(peer);
-        callback.drop_handle();
-    }
+class CallbackData {
+    //This is no op and does not need to call delete_handle since ~CallbackData is always called by the RealmNotificationTokenHandle finalizer
+    static void finalize_handle(void* isolate_callback_data, void* peer) {}
 
-    void drop_handle()
-    {
-        if (handle_ != nullptr)
-        {
-            Dart_DeleteWeakPersistentHandle_DL(handle_);
-            handle_ = nullptr;
+    void delete_handle() {
+        if (m_handle) {
+            //TODO: uncomment when the HACK is removed.
+            //Dart_DeleteWeakPersistentHandle_DL(m_handle);
+            m_handle = nullptr;
         }
     }
 
 public:
-    Callback(Dart_Handle handle, realm_dart_on_collection_change_func_t callback)
-        : handle_(Dart_NewWeakPersistentHandle_DL(handle, this, 1 << 24, finalize_)), callback_(callback) {}
+    CallbackData(Dart_Handle handle, realm_dart_on_collection_change_func_t callback)
+        : m_handle(Dart_NewFinalizableHandle_DL(handle, nullptr, 1, finalize_handle)), m_callback(callback) 
+    {}
 
-    ~Callback()
-    {
-        drop_handle();
+    ~CallbackData() {
+        delete_handle();
     }
 
-    void operator()(const realm_collection_changes_t *changes)
-    {
-        if (handle_ != nullptr)
-        {
-            Scope s; // ensure we can create handles
-            auto h = Dart_HandleFromWeakPersistent_DL(handle_);
-            // Note Dart_IsNull(h) is not exposed in DL, so we cannot check.
-            // Hence the callback has to be prepared for this eventuality.
-            callback_(h, changes);
+    void callback(const realm_collection_changes_t* changes) {
+        if (m_handle) {
+            HandleScope scope;
+            //TODO: HACK. We can not release Dart persitent handles in delete_handle on Isolate teardown since the IsolateGroup is destroyed before it.
+            //This works since Dart_WeakPersistentHandle is equivalent to Dart_FinalizableHandle. They both are FinalizablePersistentHandle internally.
+            Dart_WeakPersistentHandle weakHnd = reinterpret_cast<Dart_WeakPersistentHandle>(m_handle);
```

Comment:
This is not safe. `m_handle` is really an `Dart_FinalizableHandle` which will be auto deleted. We do not know that at this point that it is not (Depends on the order the finalizers for the notification_token_t and m_handle).

Also, this depends on an implementation detail in the dart VM. Google can change this without even bumping the minor version, and still be compliant with semantic versioning.. not that I think they are highly likely to change this soon.